### PR TITLE
feat(write_labels): store all sparse X  and raw.X in canonical form

### PR DESCRIPTION
**Pull request by Bento007**
When writing labels if the sparse matrix support canonical form, then it is converted to canonical before writing the label h5ad file out. A test has been added to verify the out of write labels is canonical.

This PR is a result of the fix in https://github.com/chanzuckerberg/single-cell-curation/pull/598 failing to resolve the seurat value cause by non-canonical matrix. https://github.com/chanzuckerberg/single-cell/issues/519 the original bug.

This was test in rdev.  Here are the steps:
1. create an rdev environment off of portal main.
2. Replace the `cellxgene-schema` requirement in requirement-processing.txt with `cellxgene-schema@git+https://github.com/chanzuckerberg/single-cell-curation.git@8c57ca36fe648842fd1128af84a58fcce1390d73#subdirectory=cellxgene_schema_cli`
3. Download https://corpora-data-prod.s3.us-west-2.amazonaws.com/b38bf00a-d413-4823-8766-fc411c61d35a/raw.h5ad
4. migrate the dataset to 3.1.0 `$ cellxgene-schema migrate raw.h5ad migrated.h5ad`
5. create a collection and submit the migrated.h5ad.
6. After around 2 hours the collection will be updated with the new dataset. 

For the unit-tests a [COO_matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html) is used. It allows creating noncanocal matrixes.

### CODE REVIEW

**Review by nayib-jose-gloria — CHANGES_REQUESTED** [review]
with this fix, can we safely revert the previous change to read the adata into memory in the write-labels step? 

Also, for the test, can we try introducing a small h5ad fixture that does not have canonical format already and running it in test_validate? Either by slightly changing the existing small test fixture or perhaps subsetting the erroneous dataset

File: cellxgene_schema_cli/cellxgene_schema/utils.py [hunk 1]
```diff
@@ -122,3 +122,17 @@ def read_h5ad(h5ad_path: Union[str, bytes, os.PathLike]) -> ad.AnnData:
         sys.exit(1)
 
     return adata
+
+
+def enforce_canonical_format(adata: ad.AnnData):
+    def _enforce_canonical_format(df):
+        X = df.X
+        canonical = getattr(X, "has_canonical_format", None)
+        if canonical is False:
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/cellxgene_schema/utils.py:131]: nit: should this be canonical is not True? In case it is None

**Comment by Bento007** [comment]
Update the rdev with the changes and ran submissions. The run is still clean https://memory-frontend.rdev.single-cell.czi.technology/collections/e5f6eb8f-422a-4e85-b249-e17fb71dd5dc

File: cellxgene_schema_cli/tests/test_utils.py [hunk 2]
```diff
@@ -71,3 +79,31 @@ def test_map_ontology_term__(adata_without_raw):
...
+class TestEnforceCanonical:
+    def test_adata_with_noncanonical_X_and_raw_X(self, noncanonical_matrix):
+        assert noncanonical_matrix.has_canonical_format is False
+        adata = AnnData(noncanonical_matrix)
+        enforce_canonical_format(adata)
+        assert adata.X.has_canonical_format is True
+
+    def test_adata_with_noncanonical_raw_X(self, noncanonical_matrix):
+        assert noncanonical_matrix.has_canonical_format is False
+        adata = AnnData(raw=AnnData(noncanonical_matrix))
```
Comment (atolopko-czi) [inline cellxgene_schema_cli/tests/test_utils.py:99]: interesting, you can pass an entire AnnData as `raw` param? Same as `raw=AnnData(noncanonical_matrix).X`?

**Review by atolopko-czi — CHANGES_REQUESTED** [review]
One question before approving, and a few nits for good measure. :)

File: cellxgene_schema_cli/cellxgene_schema/utils.py [hunk 3]
```diff
@@ -122,3 +122,23 @@ def read_h5ad(h5ad_path: Union[str, bytes, os.PathLike]) -> ad.AnnData:
         sys.exit(1)
 
     return adata
+
+
+def enforce_canonical_format(adata: ad.AnnData):
+    """
+    Enforce canonical format for anndata X and raw.X. All operation are done inplace.
```
Comment (atolopko-czi) [inline cellxgene_schema_cli/cellxgene_schema/utils.py:129]: worth nothing *why* canonicalization is needed (Seurat)

File: cellxgene_schema_cli/cellxgene_schema/utils.py [hunk 4]
```diff
@@ -122,3 +122,23 @@ def read_h5ad(h5ad_path: Union[str, bytes, os.PathLike]) -> ad.AnnData:
...
+    :param adata:
+    :return:
+    """
+
+    def _enforce_canonical_format(df):
+        X = df.X
+        canonical = getattr(X, "has_canonical_format", None)
+        if canonical is False:
+            X.sum_duplicates()
```
[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/utils.py:138]: We previously (cellxgene-schema 3.0.1 and below) accepted non-canonical datasets. This regression was introduced in 3.0.2 and is now affecting our ability to update a dataset we previously were able to upload and convert successfully to Seurat. From my understanding, product is still deciding whether to add this as a schema constraint in the future, but in the near-term we want to achieve parity with the 3.0.1 behavior of accepting non-canonical uploads.

[the remark this answers is not in the document]
Comment (atolopko-czi) [reply cellxgene_schema_cli/cellxgene_schema/utils.py:138]: I suppose it depends upon how downstream processes _use_ a non-canonical matrix and whether that processing  effectively sums the dups or (erroneously?) ignores the dups. It seems sensible to sum at any rate, but just wanted to bring it up as a potential concern.  Good to know that future schema constraints may address this properly.

**Review by atolopko-czi — APPROVED** [review]
LGTM

**Review by nayib-jose-gloria — CHANGES_REQUESTED** [review]
Note: Let's run bumpversion before merging, to release as patch 3.1.3

File: cellxgene_schema_cli/tests/test_utils.py (hunk 2 above)
Comment (Bento007) [reply cellxgene_schema_cli/tests/test_utils.py:99]: you can 😄  https://anndata.readthedocs.io/en/0.7.8/generated/anndata.AnnData.raw.html#anndata.AnnData.raw

**Review by nayib-jose-gloria — APPROVED** [review]
Tested with CSR, CSC matrices in rdev as well as the non-canonical dataset that's been causing issues and its looking good

File: cellxgene_schema_cli/cellxgene_schema/utils.py [hunk 5]
```diff
@@ -135,10 +135,12 @@ def _enforce_canonical_format(df):
         X = df.X
         if hasattr(X, "has_canonical_format") and not X.has_canonical_format:
             # this enforces canonical form; see https://docs.scipy.org/doc/scipy/tutorial/sparse.html#canonical-formats
+            logger.warning("noncanonical data found in X; converting to canonical format.")
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/cellxgene_schema/utils.py:138]: We should explicitly say we're converting to canonical by summing duplicates. In the spirit of capturing Bruce's concern that perhaps the contributor did not intend to sum the duplicate entries

**Pull request merged**